### PR TITLE
BellatrixProject

### PR DESCRIPTION
**Pull request by IE-Georgieva**

### CODE REVIEW

File: src/main/java/pageObject/ContactFormElements.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    private WebElement goldAccessPass = SearchContext.findElement(By.cssSelector("#wpforms-3347-field_3_3"));
+    private WebElement firstSession = SearchContext.findElement(By.cssSelector("#wpforms-3347-field_4_1"));
+    private WebElement secondSession = SearchContext.findElement(By.cssSelector("#wpforms-3347-field_4_2"));
+    private WebElement thirdSession = SearchContext.findElement(By.cssSelector("#wpforms-3347-field_4_3"));
+    private WebElement fourthSession = SearchContext.findElement(By.cssSelector("#wpforms-3347-field_4_4"));
+    private WebElement fifthSession = SearchContext.findElement(By.cssSelector("#wpforms-3347-field_4_5"));
+    private WebElement yesStayingOvernightRadioButton = SearchContext.findElement(By.cssSelector("#wpforms-3347-field_5_1"));
+    private WebElement noStayingOvernightRadioButton = SearchContext.findElement(By.cssSelector("#wpforms-3347-field_5_2"));
+    private WebElement checkboxReceiveEmail = SearchContext.findElement(By.cssSelector("#wpforms-3347-field_6_1"));
+    private WebElement commentsOrQuestionsSection = SearchContext.findElement(By.cssSelector("#wpforms-3347-field_7"));
```
Comment (te4eto) [inline src/main/java/pageObject/ContactFormElements.java:25]: locators are really bad

File: src/main/java/pageObject/HomepageElements.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package pageObject;
+
+import org.openqa.selenium.WebElement;
+
+public class HomepageElements extends PageObject {
+    public HomepageElements(WebElement searchContext) {
+        super(searchContext);
+    }
+
```
Comment (te4eto) [inline src/main/java/pageObject/HomepageElements.java:9]: Why is this class empty?

File: src/main/java/pageObject/OutOfStock.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package pageObject;
+
+import org.openqa.selenium.WebElement;
+
```
Comment (te4eto) [inline src/main/java/pageObject/OutOfStock.java:4]: Why is this class empty?

File: src/main/java/pageObject/ContactFormElements.java (hunk 1 above)
Comment (te4eto) [reply src/main/java/pageObject/ContactFormElements.java:25]: Are you using this anywhere?

File: src/main/java/pages/CheckOutPage.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        webDriverWait = new WebDriverWait(webDriver, Duration.ofSeconds(15));
+        webDriverWait.until(ExpectedConditions.visibilityOf(firstName));
+        firstName.sendKeys(first);
+        lastName.sendKeys(last);
+        streetAddress.sendKeys(address);
+        city.sendKeys(town);
+        postcode.sendKeys(zip);
+        phone.sendKeys(phoneNumber);
+        email.sendKeys(emailAddress);
+
```
Comment (te4eto) [inline src/main/java/pages/CheckOutPage.java:50]: remove empty row

File: src/main/java/pages/CheckOutPage.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        var element = webDriver.findElement(By.cssSelector("div.form-row"));
+        actions.moveToElement(element).moveByOffset(0, -25).click().perform();
+    }
+
+    public void loadingIndicator() {
+        performWait();
+        webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("div.blockOverlay")));
+        webDriverWait.until(ExpectedConditions.invisibilityOfElementLocated(By.cssSelector("div.blockOverlay")));
+    }
+
```
Comment (te4eto) [inline src/main/java/pages/CheckOutPage.java:65]: empty row

File: src/main/java/pages/CommonPage.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
+package pages;
+
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+import java.time.Duration;
+
```
Comment (te4eto) [inline src/main/java/pages/CommonPage.java:9]: remove empty rows in the class

File: src/main/java/pages/Header.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import pageObject.HeaderElements;
+import pageObject.PageObject;
+
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Header extends PageObject {
+
+    private List<WebElement> allMenus = SearchContext.findElements(By.cssSelector("ul.nav-menu li"));
```
Comment (te4eto) [inline src/main/java/pages/Header.java:15]: change the name of the var(e.g. navigationMenuButtons)

File: src/main/java/pages/CheckOutPage.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+import org.openqa.selenium.interactions.Actions;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+import org.testng.Assert;
+import pageObject.PageObject;
+
+import java.time.Duration;
+
+public class CheckOutPage extends CommonPage {
+
```
Comment (te4eto) [inline src/main/java/pages/CheckOutPage.java:13]: empty rows

File: src/main/java/pages/ReadMore.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    private WebElement productTitle = webDriver.findElement(By.cssSelector(".product_title.entry-title"));
+
+    public ReadMore(WebDriver driver) {
+        super(driver);
+    }
+
+    public void verifyThatButtonIsClickedSuccessfully(String expectedText, String messageOnFailure) {
+        String actualText = productTitle.getText();
+        Assert.assertTrue(actualText.contains(expectedText), messageOnFailure);
+
```
Comment (te4eto) [inline src/main/java/pages/ReadMore.java:19]: empty rows

File: src/main/java/pages/OrderReceived.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        super(driver);
+    }
+
+    public void verifyThatOrderIsReceivedSuccessfully(String expectedText, String messageOnFailure) {
+        performWait();
+        webDriverWait.until(ExpectedConditions.visibilityOf(orderReceivedText));
+        String actualText = orderReceivedText.getText();
+        Assert.assertTrue(actualText.contains(expectedText), messageOnFailure);
+    }
+
```
Comment (te4eto) [inline src/main/java/pages/OrderReceived.java:28]: rows

File: src/main/java/pages/HomePage.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import pageObject.InStockProduct;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class HomePage extends CommonPage {
```
Comment (te4eto) [inline src/main/java/pages/HomePage.java:11]: you have instock/outofstock products but you are not using outofstock

File: src/main/java/pages/Header.java [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+            if (menu.getMenuTitle().equals(menuName)) {
+                resultMenu = menu;
+                break;
+            }
+        }
+
+        return resultMenu;
+    }
+
```
Comment (te4eto) [inline src/main/java/pages/Header.java:45]: rows

File: src/main/java/pages/CommonPage.java [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+        return new Header(header);
+    }
+
+    protected WebDriverWait webDriverWait;
+
+    public void performWait() {
+        webDriverWait = new WebDriverWait(webDriver, Duration.ofSeconds(15));
+
+    }
+
```
Comment (te4eto) [inline src/main/java/pages/CommonPage.java:27]: rows

File: src/test/java/Tests.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+import java.time.Duration;
+import java.util.List;
+
+public class Tests extends BaseTest {
+
+    @Test
+    public void successfullyAddToCartElement() throws InterruptedException {
+        InStockProduct firstElement = homepage.getProductByName("Falcon 9");
+        firstElement.clickOnButton();
+        Thread.sleep(2000);
```
Comment (te4eto) [inline src/test/java/Tests.java:19]: You can look into implicit/explicit waits to avoid Thread sleeps

File: src/test/java/Tests.java [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+import java.time.Duration;
+import java.util.List;
+
+public class Tests extends BaseTest {
+
+    @Test
+    public void successfullyAddToCartElement() throws InterruptedException {
+        InStockProduct firstElement = homepage.getProductByName("Falcon 9");
+        firstElement.clickOnButton();
```
Comment (te4eto) [inline src/test/java/Tests.java:18]: this can become:
homepage.getProductByName("Falcon 9").clickOnButton();

File: src/test/java/Tests.java [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+        Thread.sleep(2000);
+        HeaderElements thirdElement = homepage.getHeader().getMenuByName("Cart");
+        thirdElement.clickOnMenu();
```
Comment (te4eto) [inline src/test/java/Tests.java:21]: same as the product:
homepage.getHeader().getMenuByName("Cart").clickOnMenu();

File: src/test/java/Tests.java [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+        cart = new Cart(driver);
+        cart.verifyThatElementIsAddedSuccessfully();
+
```
Comment (te4eto) [inline src/test/java/Tests.java:24]: row

File: src/test/java/Tests.java [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Test
+    public void successfullyClickOnReadMoreButton() throws InterruptedException {
+        InStockProduct thirdElement = homepage.getProductByName("Proton-M");
```
Comment (te4eto) [inline src/test/java/Tests.java:29]: Same as the above. Either do it without saving the element as a variable, or name the variables better. "thirdElement" is a bad name for a product.

File: src/test/java/Tests.java [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+        thirdElement.clickOnButton();
+        readMore = new ReadMore(driver);
```
Comment (te4eto) [inline src/test/java/Tests.java:31]: add "Page" suffix to all pages.

Comment (te4eto) [reply src/test/java/Tests.java:31]: follow the convention that you have started

File: src/test/java/Tests.java [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+        readMore.verifyThatButtonIsClickedSuccessfully("Proton-M", "Read more button is not working correctly!");
+
+    }
+
+    @Test
+    public void successfullyOrder() throws InterruptedException {
```
Comment (te4eto) [inline src/test/java/Tests.java:37]: follow all of the above comments to fix this test

File: src/test/java/Tests.java [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+        firstElement.clickOnButton();
+        Thread.sleep(2000);
+        HeaderElements thirdElement = homepage.getHeader().getMenuByName("Cart");
+        thirdElement.clickOnMenu();
+        cart = new Cart(driver);
+        cart.clickOnProceedToCheckoutButton();
+        checkOutPage = new CheckOutPage(driver);
+        checkOutPage.fillCheckoutPage("TestFirstName", "TestLastName", "Mladost 1, 5451 A", "Sofia", "1000", "02365154210", "testEmail@abv.bg");
+        checkOutPage.clickOnPlaceOrderButton();
+        Thread.sleep(10000);
```
Comment (te4eto) [inline src/test/java/Tests.java:48]: Do we need this big thread sleep here? Wouldn't implicit wait be better?

File: src/test/java/Tests.java [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+        orderReceived = new OrderReceived(driver);
+        orderReceived.verifyThatOrderIsReceivedSuccessfully("Thank you. Your order has been received.", "Order is not received.");
+    }
+
```
Comment (te4eto) [inline src/test/java/Tests.java:52]: row

File: src/main/java/pages/CheckOutPage.java (hunk 8 above)
Comment (IE-Georgieva) [reply src/main/java/pages/CheckOutPage.java:13]: Empty rows are deleted.

File: src/main/java/pages/CommonPage.java (hunk 6 above)
Comment (IE-Georgieva) [reply src/main/java/pages/CommonPage.java:9]: Empty rows are deleted.

File: src/main/java/pages/CommonPage.java (hunk 13 above)
Comment (IE-Georgieva) [reply src/main/java/pages/CommonPage.java:27]: Empty rows are deleted.

File: src/main/java/pages/Header.java (hunk 7 above)
Comment (IE-Georgieva) [reply src/main/java/pages/Header.java:15]: Name of the var is changed.

File: src/main/java/pages/Header.java (hunk 12 above)
Comment (IE-Georgieva) [reply src/main/java/pages/Header.java:45]: Empty rows are deleted.

File: src/main/java/pages/HomePage.java (hunk 11 above)
Comment (IE-Georgieva) [reply src/main/java/pages/HomePage.java:11]: Outofstock class is deleted.

File: src/main/java/pages/OrderReceived.java (hunk 10 above)
Comment (IE-Georgieva) [reply src/main/java/pages/OrderReceived.java:28]: Empty rows are deleted.

File: src/main/java/pages/ReadMore.java (hunk 9 above)
Comment (IE-Georgieva) [reply src/main/java/pages/ReadMore.java:19]: Empty rows are deleted.

File: src/test/java/Tests.java (hunk 15 above)
Comment (IE-Georgieva) [reply src/test/java/Tests.java:18]: ready

File: src/test/java/Tests.java (hunk 17 above)
Comment (IE-Georgieva) [reply src/test/java/Tests.java:24]: Empty rows are deleted.

File: src/test/java/Tests.java (hunk 21 above)
Comment (IE-Georgieva) [reply src/test/java/Tests.java:48]: Thread sleep is changed with implicit wait

File: src/test/java/Tests.java (hunk 22 above)
Comment (IE-Georgieva) [reply src/test/java/Tests.java:52]: Empty row is deleted

**Pull request merged**